File: client/Scripts/messages/messages.cs
using System;
using System.IO;
using System.Text;

public class LoginMessage {
	public string Name
	public string Password

	public Serialize(BinaryWriter writer) {
		writer.Write(this.Name.length);
		writer.Write(this.Name);
		writer.Write(this.Password.length);
		writer.Write(this.Password);
	}

	public void Deserialize(BinaryReader buffer) {
		var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	temp0_1 := make([]byte, l0_1)
	buffer.Read(temp0_1)
	this.Name = string(temp0_1)
		var l1_1 int
	binary.Read(buffer, binary.LittleEndian, &l1_1)
	temp1_1 := make([]byte, l1_1)
	buffer.Read(temp1_1)
	this.Password = string(temp1_1)
	}
}

public class ListGames {

	public Serialize(BinaryWriter writer) {
	}

	public void Deserialize(BinaryReader buffer) {
	}
}

public class ListGamesResponse {
	public []uint32 IDs
	public []string Names

	public Serialize(BinaryWriter writer) {
		writer.Write(this.IDs.length);
		for ( int i=0; i < this.IDs.length; i++) {
			writer.Write(this.IDs[i]);
		}
		writer.Write(this.Names.length);
		for ( int i=0; i < this.Names.length; i++) {
			writer.Write(this.Names[i].length);
			writer.Write(this.Names[i]);
		}
	}

	public void Deserialize(BinaryReader buffer) {
		var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	this.IDs = make([]uint32, l0_1)
	for i := 0; i < l0_1; i++ {
			this.IDs[i] = buffer.ReadUInt32();
	}
		var l1_1 int
	binary.Read(buffer, binary.LittleEndian, &l1_1)
	this.Names = make([]string, l1_1)
	for i := 0; i < l1_1; i++ {
			var l0_2 int
		binary.Read(buffer, binary.LittleEndian, &l0_2)
		temp0_2 := make([]byte, l0_2)
		buffer.Read(temp0_2)
		this.Names[i] = string(temp0_2)
	}
	}
}

public class CreateGame {
	public string Name

	public Serialize(BinaryWriter writer) {
		writer.Write(this.Name.length);
		writer.Write(this.Name);
	}

	public void Deserialize(BinaryReader buffer) {
		var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	temp0_1 := make([]byte, l0_1)
	buffer.Read(temp0_1)
	this.Name = string(temp0_1)
	}
}

public class JoinGame {
	public uint32 ID
	public uint32 CharID

	public Serialize(BinaryWriter writer) {
		writer.Write(this.ID);
		writer.Write(this.CharID);
	}

	public void Deserialize(BinaryReader buffer) {
		this.ID = buffer.ReadUInt32();
		this.CharID = buffer.ReadUInt32();
	}
}

public class CreateCharacter {
	public string Name
	public byte Kit

	public Serialize(BinaryWriter writer) {
		writer.Write(this.Name.length);
		writer.Write(this.Name);
		writer.Write(this.Kit);
	}

	public void Deserialize(BinaryReader buffer) {
		var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	temp0_1 := make([]byte, l0_1)
	buffer.Read(temp0_1)
	this.Name = string(temp0_1)
		this.Kit = buffer.ReadByte()
	}
}

public class DeleteCharacter {
	public int32 ID

	public Serialize(BinaryWriter writer) {
		writer.Write(this.ID);
	}

	public void Deserialize(BinaryReader buffer) {
		this.ID = buffer.ReadInt32();
	}
}

public class MapLoaded {
	public [][]byte Tiles
	public []*Entity Entities

	public Serialize(BinaryWriter writer) {
		writer.Write(this.Tiles.length);
		for ( int i=0; i < this.Tiles.length; i++) {
			writer.Write(this.Tiles[i].length);
			for ( int i=0; i < this.Tiles[i].length; i++) {
				writer.Write(this.Tiles[i][i]);
			}
		}
		writer.Write(this.Entities.length);
		for ( int i=0; i < this.Entities.length; i++) {
			this.Entities[i].Serialize(buffer)
		}
	}

	public void Deserialize(BinaryReader buffer) {
		var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	this.Tiles = make([][]byte, l0_1)
	for i := 0; i < l0_1; i++ {
			var l0_2 int
		binary.Read(buffer, binary.LittleEndian, &l0_2)
		this.Tiles[i] = make([]byte, l0_2)
		for i := 0; i < l0_2; i++ {
				this.Tiles[i][i] = buffer.ReadByte()
		}
	}
		var l1_1 int
	binary.Read(buffer, binary.LittleEndian, &l1_1)
	this.Entities = make([]*Entity, l1_1)
	for i := 0; i < l1_1; i++ {
			this.Entities[i] = new(Entity)
		this.Entities[i].Deserialize(buffer)
	}
	}
}

public class Entity {
	public uint32 ID
	public byte HealthPercent
	public int32 X
	public int32 Y

	public Serialize(BinaryWriter writer) {
		writer.Write(this.ID);
		writer.Write(this.HealthPercent);
		writer.Write(this.X);
		writer.Write(this.Y);
	}

	public void Deserialize(BinaryReader buffer) {
		this.ID = buffer.ReadUInt32();
		this.HealthPercent = buffer.ReadByte()
		this.X = buffer.ReadInt32();
		this.Y = buffer.ReadInt32();
	}
}

public class EntityMove {
	public byte Direction

	public Serialize(BinaryWriter writer) {
		writer.Write(this.Direction);
	}

	public void Deserialize(BinaryReader buffer) {
		this.Direction = buffer.ReadByte()
	}
}

public class UseAbility {
	public int32 AbilityID
	public uint32 Target

	public Serialize(BinaryWriter writer) {
		writer.Write(this.AbilityID);
		writer.Write(this.Target);
	}

	public void Deserialize(BinaryReader buffer) {
		this.AbilityID = buffer.ReadInt32();
		this.Target = buffer.ReadUInt32();
	}
}

public class AbilityResult {
	public *Entity Target
	public int32 Damage
	public byte State

	public Serialize(BinaryWriter writer) {
		this.Target.Serialize(buffer)
		writer.Write(this.Damage);
		writer.Write(this.State);
	}

	public void Deserialize(BinaryReader buffer) {
		this.Target = new(Entity)
	this.Target.Deserialize(buffer)
		this.Damage = buffer.ReadInt32();
		this.State = buffer.ReadByte()
	}
}

public class EndGame {

	public Serialize(BinaryWriter writer) {
	}

	public void Deserialize(BinaryReader buffer) {
	}
}

File: server/messages/messages.go
package messages

import (
	"bytes"
	"encoding/binary"
)

type LoginMessage struct {
	Name string
	Password string
}

func (m *LoginMessage) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, len(m.Name))
	buffer.WriteString(m.Name)
	binary.Write(buffer, binary.LittleEndian, len(m.Password))
	buffer.WriteString(m.Password)
}

func (m *LoginMessage) Deserialize(buffer *bytes.Buffer) {
	var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	temp0_1 := make([]byte, l0_1)
	buffer.Read(temp0_1)
	m.Name = string(temp0_1)
	var l1_1 int
	binary.Read(buffer, binary.LittleEndian, &l1_1)
	temp1_1 := make([]byte, l1_1)
	buffer.Read(temp1_1)
	m.Password = string(temp1_1)
}

type ListGames struct {
}

func (m *ListGames) Serialize(buffer *bytes.Buffer) {
}

func (m *ListGames) Deserialize(buffer *bytes.Buffer) {
}

type ListGamesResponse struct {
	IDs []uint32
	Names []string
}

func (m *ListGamesResponse) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, len(m.IDs))
	for _, v2 := range m.IDs {
		binary.Write(buffer, binary.LittleEndian, v2)
	}
	binary.Write(buffer, binary.LittleEndian, len(m.Names))
	for _, v2 := range m.Names {
		binary.Write(buffer, binary.LittleEndian, len(v2))
		buffer.WriteString(v2)
	}
}

func (m *ListGamesResponse) Deserialize(buffer *bytes.Buffer) {
	var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	m.IDs = make([]uint32, l0_1)
	for i := 0; i < l0_1; i++ {
		binary.Read(buffer, binary.LittleEndian, &m.IDs[i])
	}
	var l1_1 int
	binary.Read(buffer, binary.LittleEndian, &l1_1)
	m.Names = make([]string, l1_1)
	for i := 0; i < l1_1; i++ {
		var l0_2 int
		binary.Read(buffer, binary.LittleEndian, &l0_2)
		temp0_2 := make([]byte, l0_2)
		buffer.Read(temp0_2)
		m.Names[i] = string(temp0_2)
	}
}

type CreateGame struct {
	Name string
}

func (m *CreateGame) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, len(m.Name))
	buffer.WriteString(m.Name)
}

func (m *CreateGame) Deserialize(buffer *bytes.Buffer) {
	var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	temp0_1 := make([]byte, l0_1)
	buffer.Read(temp0_1)
	m.Name = string(temp0_1)
}

type JoinGame struct {
	ID uint32
	CharID uint32
}

func (m *JoinGame) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, m.ID)
	binary.Write(buffer, binary.LittleEndian, m.CharID)
}

func (m *JoinGame) Deserialize(buffer *bytes.Buffer) {
	binary.Read(buffer, binary.LittleEndian, &m.ID)
	binary.Read(buffer, binary.LittleEndian, &m.CharID)
}

type CreateCharacter struct {
	Name string
	Kit byte
}

func (m *CreateCharacter) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, len(m.Name))
	buffer.WriteString(m.Name)
	buffer.WriteByte(m.Kit)
}

func (m *CreateCharacter) Deserialize(buffer *bytes.Buffer) {
	var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	temp0_1 := make([]byte, l0_1)
	buffer.Read(temp0_1)
	m.Name = string(temp0_1)
	m.Kit, _ = buffer.ReadByte()
}

type DeleteCharacter struct {
	ID int32
}

func (m *DeleteCharacter) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, m.ID)
}

func (m *DeleteCharacter) Deserialize(buffer *bytes.Buffer) {
	binary.Read(buffer, binary.LittleEndian, &m.ID)
}

type MapLoaded struct {
	Tiles [][]byte
	Entities []*Entity
}

func (m *MapLoaded) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, len(m.Tiles))
	for _, v2 := range m.Tiles {
		binary.Write(buffer, binary.LittleEndian, len(v2))
		for _, v3 := range v2 {
			buffer.WriteByte(v3)
		}
	}
	binary.Write(buffer, binary.LittleEndian, len(m.Entities))
	for _, v2 := range m.Entities {
		v2.Serialize(buffer)
	}
}

func (m *MapLoaded) Deserialize(buffer *bytes.Buffer) {
	var l0_1 int
	binary.Read(buffer, binary.LittleEndian, &l0_1)
	m.Tiles = make([][]byte, l0_1)
	for i := 0; i < l0_1; i++ {
		var l0_2 int
		binary.Read(buffer, binary.LittleEndian, &l0_2)
		m.Tiles[i] = make([]byte, l0_2)
		for i := 0; i < l0_2; i++ {
			m.Tiles[i][i], _ = buffer.ReadByte()
		}
	}
	var l1_1 int
	binary.Read(buffer, binary.LittleEndian, &l1_1)
	m.Entities = make([]*Entity, l1_1)
	for i := 0; i < l1_1; i++ {
		m.Entities[i] = new(Entity)
		m.Entities[i].Deserialize(buffer)
	}
}

type Entity struct {
	ID uint32
	HealthPercent byte
	X int32
	Y int32
}

func (m *Entity) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, m.ID)
	buffer.WriteByte(m.HealthPercent)
	binary.Write(buffer, binary.LittleEndian, m.X)
	binary.Write(buffer, binary.LittleEndian, m.Y)
}

func (m *Entity) Deserialize(buffer *bytes.Buffer) {
	binary.Read(buffer, binary.LittleEndian, &m.ID)
	m.HealthPercent, _ = buffer.ReadByte()
	binary.Read(buffer, binary.LittleEndian, &m.X)
	binary.Read(buffer, binary.LittleEndian, &m.Y)
}

type EntityMove struct {
	Direction byte
}

func (m *EntityMove) Serialize(buffer *bytes.Buffer) {
	buffer.WriteByte(m.Direction)
}

func (m *EntityMove) Deserialize(buffer *bytes.Buffer) {
	m.Direction, _ = buffer.ReadByte()
}

type UseAbility struct {
	AbilityID int32
	Target uint32
}

func (m *UseAbility) Serialize(buffer *bytes.Buffer) {
	binary.Write(buffer, binary.LittleEndian, m.AbilityID)
	binary.Write(buffer, binary.LittleEndian, m.Target)
}

func (m *UseAbility) Deserialize(buffer *bytes.Buffer) {
	binary.Read(buffer, binary.LittleEndian, &m.AbilityID)
	binary.Read(buffer, binary.LittleEndian, &m.Target)
}

type AbilityResult struct {
	Target *Entity
	Damage int32
	State byte
}

func (m *AbilityResult) Serialize(buffer *bytes.Buffer) {
	m.Target.Serialize(buffer)
	binary.Write(buffer, binary.LittleEndian, m.Damage)
	buffer.WriteByte(m.State)
}

func (m *AbilityResult) Deserialize(buffer *bytes.Buffer) {
	m.Target = new(Entity)
	m.Target.Deserialize(buffer)
	binary.Read(buffer, binary.LittleEndian, &m.Damage)
	m.State, _ = buffer.ReadByte()
}

type EndGame struct {
}

func (m *EndGame) Serialize(buffer *bytes.Buffer) {
}

func (m *EndGame) Deserialize(buffer *bytes.Buffer) {
}

